rust/src/delta.rs

```diff
@@ -1329,64 +1277,89 @@ impl<'a> DeltaTransaction<'a> {
+        let prepared_commit = self.prepare_commit(_operation).await?;
 
+        // try to commit in a loop in case other writers write the next version first
+        let version = self.try_commit_loop(&prepared_commit).await?;
+
+        Ok(version)
+    }
+
+    /// Low-level transaction API. Creates a temporary commit file.
+    /// Once created, the actual commit could be executed with `try_commit`.
+    pub async fn prepare_commit(
+        &self,
+        _operation: Option<DeltaOperation>,
+    ) -> Result<PreparedCommit, DeltaTransactionError> {
+        let token = Uuid::new_v4().to_string();
+
+        // TODO: create a CommitInfo action and prepend it to actions.
         // Serialize all actions that are part of this log entry.
         let log_entry = log_entry_from_actions(&self.actions)?;
 
-        // try to commit in a loop in case other writers write the next version first
-        let version = self.try_commit_loop(log_entry.as_bytes()).await?;
+        let file_name = format!("_commit_{}.json", token);
+        let uri = self
+            .delta_table
+            .storage
+            .join_path(&self.delta_table.log_uri, &file_name);
 
-        // NOTE: since we have the log entry in memory already,
-        // we could optimize this further by merging the log entry instead of updating from storage.
-        self.delta_table.update().await?;
+        self.delta_table
+            .storage
+            .put_obj(&uri, log_entry.as_bytes())
+            .await?;
 
-        Ok(version)
+        Ok(PreparedCommit { uri })
     }
 
-    /// Commits the delta transaction at the specified version.
-    /// Propagates version conflict errors back to the caller immediately.
-    pub async fn commit_version(
+    /// Low-level transaction API. Try to commit a prepared commit file.
+    /// Returns `DeltaTransactionError::VersionAlreadyExists` if the given `version`
+    /// already exists. The caller should handle the retry logic itself. Otherwise, if there's no
+    /// additional behavior required then one might rely on `commit` which handles the commit loop itself.
+    pub async fn try_commit(
         &mut self,
+        commit: &PreparedCommit,
         version: DeltaDataTypeVersion,
-        _operation: Option<DeltaOperation>,
     ) -> Result<DeltaDataTypeVersion, DeltaTransactionError> {
-        // TODO: create a CommitInfo action and prepend it to actions.
-
-        let log_entry = log_entry_from_actions(&self.actions)?;
-        let tmp_commit_uri = self.prepare_commit(log_entry.as_bytes()).await?;
-        let version = self.try_commit(&tmp_commit_uri, version).await?;
+        // move temporary commit file to delta log directory
+        // rely on storage to fail if the file already exists -
+        self.delta_table
+            .storage
+            .rename_obj(
+                &commit.uri,
+                &self.delta_table.commit_uri_from_version(version),
+            )
+            .await?;
```

Comment:
In the cases where an error is not retry-able or recoverable, if this operation fails then the temporary file gets left on the file system in perpetuity correct?

I don't see any problem with that at this point, but if that's the case, I think the Rustdocs should be updated to document which cases might result in a temporary commit file orphaned in storage, and the types of file names (presuming `_commit_blah.json` is a temp file name) that an admin could then go in later and clean up

Reply:
is that going to work for us? https://github.com/delta-io/delta-rs/pull/305/files#diff-3b5c80af9e5527fbeaa6dad151a835f0733a26e61d0ba0471c58f243a29dfa99R1152-R1153